diff --git a/captcha.module b/captcha.module
index 19744a1..0ff5179 100644
--- a/captcha.module
+++ b/captcha.module
@@ -365,6 +365,14 @@ function captcha_form_alter(&$form, $form_state, $form_id) {
         '#type' => 'captcha',
         '#captcha_type' => $captcha_point->module .'/'. $captcha_point->captcha_type,
       );
+   $captcha_element['captcha_query'] = array(
+      '#type' => 'value',
+      '#value' => $captcha['query']
+    );
+    $captcha_element['captcha_type'] = array(
+      '#type' => 'value',
+      '#value' => $captcha['type']
+    );
       // Add a CAPTCHA description if required.
       if (variable_get('captcha_add_captcha_description', TRUE)) {
         $captcha_element['#description'] = _captcha_get_description();
@@ -705,7 +713,10 @@ function captcha_captcha($op, $captcha_type = '') {
         $answer = mt_rand(1, 20);
         $x = mt_rand(1, $answer);
         $y = $answer - $x;
+        $query = t('@x + @y = ', array('@x' => $x, '@y' => $y));
         $result['solution'] = "$answer";
+        $result['query'] = $query;
+        $result['type'] = 'math';
         // Build challenge widget.
         // Note that we also use t() for the math challenge itself. This makes
         // it possible to 'rephrase' the challenge a bit through localization
@@ -714,7 +725,7 @@ function captcha_captcha($op, $captcha_type = '') {
           '#type' => 'textfield',
           '#title' => t('Math question'),
           '#description' => t('Solve this simple math problem and enter the result. E.g. for 1+3, enter 4.'),
-          '#field_prefix' => t('@x + @y = ', array('@x' => $x, '@y' => $y)),
+          '#field_prefix' => $query,
           '#size' => 4,
           '#maxlength' => 2,
           '#required' => TRUE,
